src/realm/impl/transact_log.hpp

```diff
@@ -793,119 +838,150 @@ inline char* TransactLogEncoder::encode_double(char* ptr, double value)
+                                                                         TransactLogEncoder::IntegerList list)
+{
+    auto i = std::get<0>(list);
+    auto end = std::get<1>(list);
+
+    while (end - i > max_numbers_per_chunk) {
+        for (int j = 0; j < max_numbers_per_chunk; ++j)
+            ptr = encode_int(ptr, *i++);
+        advance(ptr);
+        size_t max_required_bytes_2 = max_enc_bytes_per_num * max_numbers_per_chunk;
+        ptr = reserve(max_required_bytes_2); // Throws
     }
-};
+
+    while (i != end)
+        ptr = encode_int(ptr, *i++);
+
+    return ptr;
+}
+
 template <>
-struct TransactLogEncoder::EncodeNumber<double> {
-    void operator()(double value, char** ptr)
-    {
-        *ptr = encode_double(*ptr, value);
-    }
-};
+inline char* TransactLogEncoder::encode<TransactLogEncoder::UnsignedList>(char* ptr,
+                                                                          TransactLogEncoder::UnsignedList list)
+{
+    auto i = std::get<0>(list);
+    auto end = std::get<1>(list);
+
+    while (i != end)
+        ptr = encode_int(ptr, *i++);
+
+    return ptr;
+}
+
+template <class T>
+size_t TransactLogEncoder::max_size(T)
+{
+    return max_enc_bytes_per_num;
+}
+
 template <>
-struct TransactLogEncoder::EncodeNumber<DataType> {
-    void operator()(DataType type, char** ptr)
-    {
-        auto value_2 = type + 0; // Perform integral promotion
-        *ptr = encode_int(*ptr, value_2);
-    }
-};
+inline size_t TransactLogEncoder::max_size(char)
+{
+    return 1;
+}
 
-template <class L>
-void TransactLogEncoder::append_simple_instr(Instruction instr, const util::Tuple<L>& numbers)
+template <>
+inline size_t TransactLogEncoder::max_size(bool)
 {
-    size_t num_numbers = util::TypeCount<L>::value;
-    size_t max_required_bytes = 1 + max_enc_bytes_per_num * num_numbers;
-    char* ptr = reserve(max_required_bytes); // Throws
-    *ptr++ = char(instr);
-    util::for_each<EncodeNumber>(numbers, &ptr);
-    advance(ptr);
+    return 1;
+}
+
+template <>
+inline size_t TransactLogEncoder::max_size(Instruction)
+{
+    return 1;
+}
+
+template <>
+inline size_t TransactLogEncoder::max_size(DataType)
+{
+    return 1;
+}
+
+template <>
+inline size_t TransactLogEncoder::max_size(StringData s)
+{
+    return max_enc_bytes_per_num + s.size();
+}
+
+template <>
+inline size_t TransactLogEncoder::max_size<TransactLogEncoder::IntegerList>(IntegerList)
+{
+    return max_enc_bytes_per_num * max_numbers_per_chunk;
 }
 
-template <class L>
-void TransactLogEncoder::append_string_instr(Instruction instr, const util::Tuple<L>& numbers, StringData string)
+template <>
+inline size_t TransactLogEncoder::max_size<TransactLogEncoder::UnsignedList>(UnsignedList list)
 {
-    size_t num_numbers = util::TypeCount<L>::value + 1;
-    size_t max_required_bytes = 1 + max_enc_bytes_per_num * num_numbers + string.size();
+    return max_enc_bytes_per_num * (std::get<1>(list) - std::get<0>(list));
```

Comment:
Why is this computation so much different from IntegerList version? Why is element 0 and 1 special in the list - what do they mean?